src/tacacs_plus_fmt_plug.c

```diff
@@ -0,0 +1,341 @@
+
+#include <string.h>
+
+#ifdef _OPENMP
+#include <omp.h>
+#ifndef OMP_SCALE
+#define OMP_SCALE               8  // tuned on i5-6500 CPU
+#endif
+#endif
+
+#include "formats.h"
+#include "misc.h"
+#include "common.h"
+#include "params.h"
+#include "options.h"
+#include "md5.h"
+#include "memdbg.h"
+
+#define FORMAT_LABEL            "tacacs-plus"
+#define FORMAT_NAME             "TACACS+"
+#define FORMAT_TAG              "$tacacs-plus$"
+#define TAG_LENGTH              (sizeof(FORMAT_TAG) - 1)
+#define ALGORITHM_NAME          "MD5 32/" ARCH_BITS_STR
+#define BENCHMARK_COMMENT       ""
+#define BENCHMARK_LENGTH        0
+#define PLAINTEXT_LENGTH        125
+#define BINARY_SIZE             0
+#define BINARY_ALIGN            1
+#define SALT_SIZE               sizeof(struct custom_salt)
+#define SALT_ALIGN              sizeof(uint32_t)
+#define MIN_KEYS_PER_CRYPT      1
+#define MAX_KEYS_PER_CRYPT      64  // tuned on i5-6500 CPU
+
+#define MAX_CIPHERTEXT_LENGTH   1024
+#ifndef MD5_DIGEST_LENGTH
+#define MD5_DIGEST_LENGTH       16
+#endif
+
+static struct fmt_tests tests[] = {
+	{"$tacacs-plus$0$6d0e1631$db7c01e77499$c006", "1234"},
+	{"$tacacs-plus$0$6d0e1631$d623c7692ca7b12f7ecef113bea72845$c004", "1234"},
+	{"$tacacs-plus$0$6d0e1631$f7711e4b904fc4a4753e923e9bf3d2cc33e9febd3d2db74b9aa6d20462c2072013c77345d7112400d7b915$c002", "1234"},
+	{"$tacacs-plus$0$d0cb2225$f5acae34eb560981b30315cc9b08f00a31566eb5f09351713d0b33067ea1fbdb63ed84062ec24ebc45de63$c002", "testing123"},
+	{"$tacacs-plus$0$acf4c30b$c73c409532a4a80e58ba94391111e300$c002", "12345"},
+	{NULL}
+};
+
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+static int *saved_len;
+static int *cracked;
+static size_t cracked_size;
+
+static struct custom_salt {
+	uint32_t ctlen;
+	MD5_CTX pctx;
+	uint32_t pre_hash_data_len;
+	unsigned char pre_hash_data[8];
+	uint32_t hash_data_len;
+	unsigned char hash_data[8];
+	union {
+		uint64_t chunk0;
+		unsigned char buf[MAX_CIPHERTEXT_LENGTH];
+	} ciphertext;
+} *cur_salt;
+
+static void init(struct fmt_main *self)
+{
+#ifdef _OPENMP
+	int omp_t = omp_get_max_threads();
+
+	if (omp_t > 1) {
+		self->params.min_keys_per_crypt *= omp_t;
+		omp_t *= OMP_SCALE;
+		self->params.max_keys_per_crypt *= omp_t;
+	}
+#endif
+	saved_key = mem_calloc(sizeof(*saved_key), self->params.max_keys_per_crypt);
+	saved_len = mem_calloc(self->params.max_keys_per_crypt, sizeof(*saved_len));
+	cracked_size = sizeof(*cracked) * self->params.max_keys_per_crypt;
+	cracked = mem_calloc(cracked_size, 1);
+}
+
+static void done(void)
+{
+	MEM_FREE(saved_key);
+	MEM_FREE(saved_len);
+	MEM_FREE(cracked);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *ctcopy, *keeptr, *p;
+	int value, extra;
+
+	if (strncmp(ciphertext, FORMAT_TAG, TAG_LENGTH) != 0)
+		return 0;
+
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+
+	ctcopy += TAG_LENGTH;
+	if ((p = strtokm(ctcopy, "$")) == NULL) // version / type
+		goto err;
+	if (!isdec(p))
+		goto err;
+	value = atoi(p);
```

Comment:
Not specific to this format, but in general the string containing only decimal digits isn't enough to make sure `atoi()` is formally safe to use. Signed integer overflow is UB, and too many decimal digits can trigger just that. We should avoid `atoi()` altogether.